js/packages/openinference-instrumentation-llama-index/src/instrumentation.ts

```diff
@@ -87,72 +98,160 @@ export class LlamaIndexInstrumentation extends InstrumentationBase<
       return moduleExports;
     }
 
-    // eslint-disable-next-line @typescript-eslint/no-this-alias
-    const instrumentation: LlamaIndexInstrumentation = this;
-
-    type RetrieverQueryEngineQueryType =
-      typeof moduleExports.RetrieverQueryEngine.prototype.query;
-
     this._wrap(
       moduleExports.RetrieverQueryEngine.prototype,
       "query",
-      (original: RetrieverQueryEngineQueryType): any => {
-        return function patchedQuery(
-          this: unknown,
-          ...args: Parameters<RetrieverQueryEngineQueryType>
-        ) {
-          const span = instrumentation.tracer.startSpan(`Query`, {
-            kind: SpanKind.INTERNAL,
-            attributes: {
-              [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
-                OpenInferenceSpanKind.CHAIN,
-              [SemanticConventions.INPUT_VALUE]: args[0].query,
-            },
-          });
-
-          const execContext = getExecContext(span);
-
-          const execPromise = safeExecuteInTheMiddle<
-            ReturnType<RetrieverQueryEngineQueryType>
-          >(
-            () => {
-              return context.with(execContext, () => {
-                return original.apply(this, args);
-              });
-            },
-            (error) => {
-              // Push the error to the span
-              if (error) {
-                span.recordException(error);
-                span.setStatus({
-                  code: SpanStatusCode.ERROR,
-                  message: error.message,
-                });
-                span.end();
-              }
-            },
-          );
-          const wrappedPromise = execPromise.then((result) => {
-            span.setAttributes({
-              [SemanticConventions.OUTPUT_VALUE]: result.response,
-            });
-            span.end();
-            return result;
-          });
-          return context.bind(execContext, wrappedPromise);
-        };
+      (original): any => {
+        return this.patchQueryMethod(
+          original,
+          moduleExports,
+          this
+        );
       },
     );
 
-    _isOpenInferencePatched = true;
+    this._wrap(
+      moduleExports.VectorIndexRetriever.prototype,
+      "retrieve",
+      (original): any => {
+        return this.patchRetrieveMethod(
+          original,
+          moduleExports,
+          this
+        );
+      },
+    )
 
+    _isOpenInferencePatched = true;
     return moduleExports;
   }
 
   private unpatch(moduleExports: typeof llamaindex, moduleVersion?: string) {
     this._diag.debug(`Un-patching ${MODULE_NAME}@${moduleVersion}`);
     this._unwrap(moduleExports.RetrieverQueryEngine.prototype, "query");
+    this._unwrap(moduleExports.VectorIndexRetriever.prototype, "retrieve");
 
     _isOpenInferencePatched = false;
   }
+
+  private patchQueryMethod(
+    original: typeof module.RetrieverQueryEngine.prototype.query,
+    module: typeof llamaindex,
+    instrumentation: LlamaIndexInstrumentation
```

Comment:
if this is not going to be a method on the class, you would want to pass this in, if it is, you shouldn't need to pass in this since this will still refer to the class

Reply:
ahh but you would need to store `this` in a variable however, in between 141 and 142 since you lose this inside of the patchedQueryFunction, so passing it in like this could be fine. Also means you don't have to put this method in this class and could instead just pull it out which might make it easier to test

Reply:
Yeah you have to keep track of `this` for the nested method. I'm planning on moving all of these methods to a util file once I have an idea of what util functions I need